clvm-traits/src/lib.rs

```diff
@@ -27,37 +27,157 @@ pub use to_clvm::*;
 pub use wrappers::*;
 
 #[cfg(test)]
-#[cfg(feature = "derive")]
-mod tests {
+pub mod tests {
     extern crate self as clvm_traits;
 
-    use std::fmt;
+    use super::*;
+
+    #[derive(Clone)]
+    pub enum TestNode {
+        Atom(usize),
+        Pair(Box<TestNode>, Box<TestNode>),
+    }
+
+    pub struct TestAllocator {
+        atoms: Vec<Vec<u8>>,
+    }
+
+    impl TestAllocator {
+        pub fn new() -> Self {
+            TestAllocator { atoms: vec![] }
+        }
+
+        fn new_atom(&mut self, buf: &[u8]) -> TestNode {
+            let idx = self.atoms.len();
+            self.atoms.push(buf.to_vec());
+            TestNode::Atom(idx)
+        }
 
-    use clvmr::{allocator::NodePtr, serde::node_to_bytes, Allocator};
+        fn atom(&self, idx: usize) -> &[u8] {
+            self.atoms[idx].as_slice()
+        }
+    }
+
+    pub fn node_eq(a: &TestAllocator, left: &TestNode, right: &TestNode) -> bool {
+        match (left, right) {
+            (TestNode::Atom(l), TestNode::Atom(r)) => a.atom(*l) == a.atom(*r),
+            (TestNode::Pair(l1, r1), TestNode::Pair(l2, r2)) => {
+                node_eq(a, l1, l2) && node_eq(a, r1, r2)
+            }
+            (_, _) => false,
+        }
+    }
+
+    pub fn node_to_str(a: &TestAllocator, input: &TestNode) -> String {
+        match input {
+            TestNode::Atom(v) => {
+                let atom = a.atom(*v);
+                if atom.len() == 0 {
+                    "NIL".to_owned()
+                } else {
+                    format!("{}", &hex::encode(&atom))
+                }
+            }
+            TestNode::Pair(l, r) => format!("( {} {}", node_to_str(a, l), node_to_str(a, r)),
```

Comment:
This format might be easier to read with a closing brace for cons pairs, something like ({} {})

Reply:
yeah, but it's more complicated to write a parser for :)